includes/class-fontawesome-metadata-provider.php

```diff
@@ -0,0 +1,193 @@
+
+  // phpcs:ignore Generic.Commenting.DocComment.MissingShort
+	/**
+	 * @ignore
+	 */
+  protected static $instance = null;
+
+  // phpcs:ignore Generic.Commenting.DocComment.MissingShort
+	/**
+	 * @ignore
+	 */
+	protected function get( $url, $args = array() ) {
+		return wp_remote_get( $url, $args );
+  }
+
+  // phpcs:ignore Generic.Commenting.DocComment.MissingShort
+	/**
+	 * @ignore
+	 */
+	protected function build_query_url( $url, $query ) {
+		return "{$url}?{$query}";
+	}
+
+
+	/**
+	 * Returns the FontAwesome_Metadata_Provider singleton instance.
+	 *
+	 * @return FontAwesome_Metadata_Provider
+	 */
+	public static function instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+		return self::$instance;
+	}
+
+  /**
+	 * Returns an associative array indicating the status of the status of the last network
+	 * request that attempted to retrieve releases metadata, or null if no network request has
+	 * been issued during the life time of the current Singleton instance.
+	 *
+	 * The shape of an array return looks like this:
+	 * ```php
+	 * array(
+	 *   'code' => 403,
+	 *   'message' => 'some message',
+	 * )
+	 * ```
+	 *
+	 * The value of the `code` key is one of:
+	 * - `200` if successful,
+	 * - `0` if there was some code error that prevented the network request from completing
+	 * - otherwise some HTTP error code as returned by `wp_remote_get()`
+	 *
+	 * @return array|null
+	 */
+	public function get_status() {
+		return $this->status;
+  }
+
+  // phpcs:ignore Generic.Commenting.DocComment.MissingShort
+	/**
+	 * Loads all versions.
+	 *
+	 * @ignore
+	 */
+	public function get_available_versions() {
+		$query = 'query={versions}';
+		$json = $this->metadata_query($query);
+		$version_array = array();
+
+		try {
+			$versions = $json->versions;
+      foreach($versions as $key => $val) {
+        array_push( $version_array, strval( $val ) );
+      }
+
+      usort(
+        $versions,
+        function( $first, $second ) {
+          return version_compare( $second, $first );
+        }
+      );
+
+			return $versions;
+		} catch ( Exception $e ) {
+			return array(
+					'code'    => 0,
+					'message' => 'Whoops, we failed to fetch the versions.',
+			);
+		} catch ( Error $e ) {
+			return array(
+					'code'    => 0,
+					'message' => 'Whoops, we failed when trying to fetch the versions.',
+			);
+		}
+	}
+
+	// phpcs:ignore Generic.Commenting.DocComment.MissingShort
+	/**
+	 * Provides a way to query the API and return the data as parsed json
+	 * based on the passed in query string.
+	 *
+	 * @ignore
+	 */
+	public function metadata_query($query_string) {
+		$init_status = array(
+			'code'    => null,
+			'message' => '',
+		);
+
+		$args = array(
+			'headers' => array(
+				'Content-Type' => 'application/json'
+				)
+			);
+		$url = FONTAWESOME_API_URL;
+
+		try {
+			$response = $this->get( $this->build_query_url( $url, $query_string), $args );
```

Comment:
I think it's more typical to use a POST request and include the GraphQL query document in the body of that POST. At least that's what I'm more familiar with and would prefer to do here as well. I'm a little concerned about where we may run into issues with including the whole query document encoded on the URL as a GET param.

Could you have a look at using [`wp_remote_post`](https://developer.wordpress.org/reference/functions/wp_remote_post/) instead of `wp_remote_get` for the requests to the API server?

Reply:
Good call. I updated this to use wp_remote_post.